examples/nodejs/src/marlowe-object-flow.ts

```diff
@@ -0,0 +1,659 @@
+    ],
+  };
+}
+// #endregion
+
+// #region Delay Payment State
+/**
+ * The delay payment contract can be in one of the following logical states:
+ */
+type DelayPaymentState =
+  | InitialState
+  | PaymentDeposited
+  | PaymentMissed
+  | PaymentReady
+  | Closed;
+/**
+ * In the initial state the contract is waiting for the payment to be deposited
+ */
+type InitialState = {
+  type: "InitialState";
+};
+
+/**
+ * After the payment is deposited, the contract is waiting for the payment to be released
+ */
+type PaymentDeposited = {
+  type: "PaymentDeposited";
+};
+
+/**
+ * If the payment is not deposited by the deadline, the contract can be closed.
+ * NOTE: It is not necesary to close the contract, as it will consume transaction fee (but it will release
+ *       the minUTXO)
+ */
+type PaymentMissed = {
+  type: "PaymentMissed";
+};
+
+/**
+ * After the release deadline, the payment is still in the contract, and it is ready to be released.
+ */
+type PaymentReady = {
+  type: "PaymentReady";
+};
+
+type Closed = {
+  type: "Closed";
+  result: "Missed deposit" | "Payment released";
+};
+
+function printState(state: DelayPaymentState, scheme: DelayPaymentScheme) {
+  switch (state.type) {
+    case "InitialState":
+      console.log(
+        `Waiting ${scheme.payFrom.address} to deposit ${scheme.amount}`
+      );
+      break;
+    case "PaymentDeposited":
+      console.log(
+        `Payment deposited, waiting until ${scheme.releaseDeadline} to be able to release the payment`
+      );
+      break;
+    case "PaymentMissed":
+      console.log(
+        `Payment missed on ${scheme.depositDeadline}, contract can be closed to retrieve minUTXO`
+      );
+      break;
+    case "PaymentReady":
+      console.log(`Payment ready to be released`);
+      break;
+    case "Closed":
+      console.log(`Contract closed: ${state.result}`);
+      break;
+  }
+}
+
+function getState(
+  scheme: DelayPaymentScheme,
+  currentTime: Date,
+  history: SingleInputTx[]
+): DelayPaymentState {
+  if (history.length === 0) {
+    if (currentTime < scheme.depositDeadline) {
+      return { type: "InitialState" };
+    } else {
+      return { type: "PaymentMissed" };
+    }
+  } else if (history.length === 1) {
+    // If the first transaction doesn't have an input, it means it was used to advace a timeouted contract
+    if (!history[0].input) {
+      return { type: "Closed", result: "Missed deposit" };
+    }
+    if (currentTime < scheme.releaseDeadline) {
+      return { type: "PaymentDeposited" };
+    } else {
+      return { type: "PaymentReady" };
+    }
+  } else if (history.length === 2) {
+    return { type: "Closed", result: "Payment released" };
+  } else {
+    throw new Error("Wrong state/contract, too many transactions");
+  }
+}
+
+// #endregion
+
+const mkDelayPaymentTags = (schema: DelayPaymentScheme) => {
+  const tag = "DELAY_PYMNT-1";
+  const tags = {} as Tags;
+
+  tags[`${tag}-from-0`] = splitAddress(schema.payFrom)[0];
```

Comment:
if your intent is to store only metadata more than tagging you can do something like that : 
```ts
{ "myDappId"  :{ schema : schema}}
```

Reply:
This can (and will) be improved of course. There is one discussion I want to have about TemplateParameters/Scheme to add a DSL to define it and to be able to share it inside the metadata of contract creation.

I'm abusing tags here really, Ideally I should have a single tag `DELAY_PAYMENT: v1`, and then have all the scheme encoded efficiently inside the metadata. 

The tag should only help me easily find the type of contracts that I want.

Encoding the scheme in the metadata is a simple way to share the contract sources between unconnected runtime instances. The approach you suggest doesn't cope with the max size of the value. Note that I even have to split the address in half to cope with this limitation.